library.go

```diff
@@ -0,0 +1,147 @@
+	ReadOnlyMode
+)
+
+// Repository interface represents a git.Repository, with information about
+// how it was open and where is located. Also provides Transactional
+// capabilities through the method Commit.
+type Repository interface {
+	// ID returns the RepositoryID.
+	ID() RepositoryID
+	// LocationID returns the LocationID from the Location where it was retrieved.
+	LocationID() LocationID
+	// Mode returns the Mode how it was opened.
+	Mode() Mode
+	// Commit persists all the write operations done since was open, if the
+	// repository doesn't provide Transactional capabilities should return
+	// ErrNonTransactional.
+	Commit() error
+	// Close closes the repository, if the repository was opened in transactional
+	// Mode, will delete any write operation pending to be written.
+	Close() error
+	// R returns the git.Repository.
+	R() *git.Repository
+}
+
+// LibraryID represents a Library identifier.
+type LibraryID string
+
+// Library interface represents a group of different libraries and locations,
+// it allows access to any repository stored on any library or location. Also
+// allows the iteration of the libraries and locations to perform full scan
+// operations. Library is the default entrypoint for accessing the repositories,
+// should be used when the Location is not important.
+type Library interface {
+	// ID returns the LibraryID for this Library.
+	ID() LibraryID
+	// Init initializes a new Repository in a Location, the choosen Location
+	// is dependant on the implementation, if this this not supported should
+	// return ErrNotImplemented. If a repository with the given RepositoryID
+	// already exists ErrRepositoryExists is returned.
+	Init(RepositoryID) (Repository, error)
+	// Get open a repository with the given RepositoryID, it itereates all the
+	// library locations until this repository is found. If a repository with
+	// the given RepositoryID can't be found the ErrRepositoryNotExists is
+	// returned.
+	Get(RepositoryID, Mode) (Repository, error)
+	// GetOrInit open or initilizes a Repository at a Location, if this this not
+	// supported should return ErrNotImplemented. If the repository is opened
+	// this will be done in RWMode.
+	GetOrInit(RepositoryID) (Repository, error)
+	// Has returns true, the LibraryID and the LocationID if the given
+	// RepositoryID matches any repository at any location belonging to this
+	// Library.
+	Has(RepositoryID) (bool, LibraryID, LocationID, error)
+	// Repositories returns a RepositoryIterator that iterates through all
+	// the repositories contained in all Location contained in this Library.
+	Repositories(Mode) (RepositoryIterator, error)
+	// Location returns the Location with the given LocationID, if a location
+	// can't be found ErrLocationNotExists is returned.
+	Location(LocationID) (Location, error)
+	// Locations returns a LocationIterator that iterates through all locations
+	// contained in this Library.
+	Locations() (LocationIterator, error)
```

Comment:
Make sense to have `Locations(recursive bool)  (LocationIterator, error)`, in order to allow traverse all the locations from the nested libraries?

Reply:
The locations should be expensive to retrieve, anyway I guest to make it optional